### PR TITLE
Types file and ToSBP for inlining SBP frames from messages.

### PR DESCRIPTION
**Pull request by mookerji**
/cc @mfine

### CODE REVIEW

**Comment by mookerji** [comment]
Don't merge yet!

File: generator/sbpg/targets/resources/SbpTemplate.hs [hunk 1]
```diff
@@ -9,13 +9,10 @@
 -- SBP message containers.
 
 module SwiftNav.SBP
-  ( Msg (..)
```
Comment (mfine) [inline generator/sbpg/targets/resources/SbpTemplate.hs]: Are you not exposing `Msg` any more?

Comment (mfine) [inline generator/sbpg/targets/resources/SbpTemplate.hs]: and the preamble and default sender id?

File: haskell/src/SwiftNav/SBP/Acquisition.hs [hunk 2]
```diff
@@ -61,6 +62,13 @@ instance Binary MsgAcqResult where
     putFloat32le _msgAcqResult_cf
     putWord32le _msgAcqResult_sid
 
+instance ToSBP MsgAcqResult where
+  toSBP msg senderID = encoded { _msgSBPCrc = checkCrc encoded }
+    where
+      payload = toStrict $ encode msg
+      len = fromIntegral $ Data.ByteString.length payload
+      encoded = Msg msgAcqResult senderID len payload 0
+
```
Comment (mfine) [inline haskell/src/SwiftNav/SBP/Acquisition.hs]: Use the Lens, Luke. How about:

```haskell
instance ToSBP MsgAcqResult where
  toSBP msg sender = encoded .~ msgSBPCrc (checkCrc encoded)
    where
      payload = toStrict $ encode msg
      len = fromIntegral $ Data.ByteString.length payload
      encoded = Msg msgAcqResult sender len payload 0
```

**Comment by mfine** [comment]
This is looking good!

File: generator/sbpg/targets/resources/SbpTemplate.hs (hunk 1 above)
Comment (mookerji) [inline generator/sbpg/targets/resources/SbpTemplate.hs]: Fixed!

File: haskell/src/SwiftNav/SBP/TH.hs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import Data.ByteString.Lazy hiding ( ByteString )
+import Data.Word
+import Language.Haskell.TH
+import SwiftNav.SBP.Types
+
+-- | Derive ToSBP typeclass, given an SBP message type name and the
+-- name of the implemented type.
+deriveSBP :: Name -> Name -> Q [Dec]
+deriveSBP msgType name = [d|instance ToSBP $(conT name) where
+                              toSBP msg senderID = encoded { _msgSBPCrc = checkCrc encoded }
```
Comment (mookerji) [inline haskell/src/SwiftNav/SBP/TH.hs]: @mfine so I tried lens'ing here and got some error that seemed like it shouldn't be the case: 

```
src/SwiftNav/SBP/Acquisition.hs:66:3:
    Couldn't match expected type ‘Msg’ with actual type ‘s0 -> t0’
    In the expression:
      (encoded_a7u6 .~ (msgSBPCrc (checkCrc encoded_a7u6)))
    In an equation for ‘toSBP’:
        toSBP msg_a7u2 senderID_a7u3
          = (encoded_a7u6 .~ (msgSBPCrc (checkCrc encoded_a7u6)))
          where
              payload_a7u4 = (toStrict $ (encode msg_a7u2))
              len_a7u5 = (fromIntegral $ (Data.ByteString.length payload_a7u4))
              encoded_a7u6
                = Msg msgAcqResult senderID_a7u3 len_a7u5 payload_a7u4 0
```

File: generator/sbpg/targets/resources/SbpTemplate.hs (hunk 1 above)
Comment (mfine) [inline generator/sbpg/targets/resources/SbpTemplate.hs]: Still not seeing Msg, preamble, and defaultSender

Comment (mfine) [inline generator/sbpg/targets/resources/SbpTemplate.hs]: Oh, wait - its now in Types - but you're still exposing SBPMsg - unnecessary?

File: haskell/src/SwiftNav/SBP/TH.hs (hunk 3 above)
Comment (mfine) [inline haskell/src/SwiftNav/SBP/TH.hs]: how about:

```haskell
   encoded & msgSBPCrc .~ checkCrc encoded
```

File: generator/sbpg/targets/resources/sbp-template.cabal [hunk 4]
```diff
@@ -42,6 +42,7 @@ library
                      , data-binary-ieee754
                      , lens
                      , monad-loops
+                     , template-haskell
```
Comment (mfine) [inline generator/sbpg/targets/resources/sbp-template.cabal]: Also need to expose Types as modules and include the TH in other-modules

Comment (mfine) [inline generator/sbpg/targets/resources/sbp-template.cabal]: Something like:

```
 exposed-modules:     SwiftNav.CRC16
                     , SwiftNav.SBP
                     , SwiftNav.Types
                        ....
   other-modules:      SwiftNav.TH
```

Comment (mookerji) [inline generator/sbpg/targets/resources/sbp-template.cabal]: Derp, thanks.

**Pull request merged**